Vignettes

m7pr opened the pull request:
Extends teal vignettes with new reporting features in `teal`.

gogonzo reviewed (commented):
1. "Adding Support for Reporting to Custom Modules" uses old solution with `reporter` arg in the <teal modules>$server
2. "getting started with teal" has very short subsections which links to longer documentations. Reporter here adds extra 100 lines which changes look of the vignette so much. IMO describing possible values of `init(reporter)` can be shorter just bullet-points instead of sections. Maybe we should change "Adding Support for Reporting to Custom Modules" to "Creating reproducible reports" (name tbd) and describe everything related to report there?

gogonzo commented:
> @gogonzo make sense! Do we need special vignette to cover `teal_report`? Or can it be included in "Adding Support for Reporting to Custom Modules"?

We for sure need `teal_report` vignette in `teal.report`. I don't know where in teal should we mention `teal_report`. Maybe in "Include Data in Teal Application"? I wouldn't spend too much time about `teal_report` as it serves the same role as `teal_data` and also it should be possible to specify `init(data = <teal_report>)`.

m7pr commented:
Ok, I:

- simplified and shortened `teal_report Class` in `teal.reporter`
- simplified and shortened `Getting started with teal.modules.general` in `teal`
- updated `Adding Support for Reporting to Custom Modules` in `teal`
- created `Managing Reproducible Report Documents in teal` in `teal`

Left 3 TODO in `Managing Reproducible Report Documents in teal` in `teal`

m7pr commented on vignettes/managing-reproducible-report-documents-in-teal.rmd:
```diff
@@ -0,0 +1,194 @@
+custom_reporter_template$set_template(
+  # Need to return named `teal_card`
+  function(card) {
+    tcard <- c("Templated information", card)
+    attributes(tcard) <-  attributes(card)
+    tcard
```
This is nasty. `c` removes all classes and attributes. We can substitute this example to use `edit_teal_card`

gogonzo replied:
You should use `append(card, <something to add>, after = 0)` - then it doesn't loose a class because a first argument is a `card`

m7pr replied:
Yes, I think I'll go with this one

gogonzo commented on vignettes/managing-reproducible-report-documents-in-teal.rmd:
```diff
@@ -0,0 +1,195 @@
+
+## Introduction
+
+`teal` provides comprehensive reporting capabilities that allow users to generate reproducible documents from their analysis modules. This vignette covers the key components for managing report documents in `teal` applications.
```
```suggestion
`teal` provides reporting capabilities that allow app users to compose reproducible documents from their analysis modules. This vignette covers the key components for managing report documents in `teal` applications.
```

gogonzo commented on vignettes/managing-reproducible-report-documents-in-teal.rmd:
```diff
@@ -0,0 +1,195 @@
+
+`teal` provides comprehensive reporting capabilities that allow users to generate reproducible documents from their analysis modules. This vignette covers the key components for managing report documents in `teal` applications.
+
+The reporting system offers three levels of control:
+
+1. **Global reporting control** via the `reporter` parameter in `init()`
+2. **Module-level control** using `disable_report()` to exclude specific modules  
+3. **Report customization** with the `after` function to enhance report content
```
Three levels suggests that you go deeper or higher in the hierarchy. We operate here only on global/module (local) level.
```suggestion
The reporting system offers two types of control:

1. **Global reporting control** via the `reporter` parameter in `init()`
2. **Module-level control** via `after()` and `disable_report()`
```

m7pr replied:
Make sense, thanks!

gogonzo commented on vignettes/managing-reproducible-report-documents-in-teal.rmd:
```diff
@@ -0,0 +1,195 @@
+
+This vignette demonstrates how to use these features effectively to create flexible and customized reporting workflows.
+
+## Global and Local Reporting Control
```
I think it's worth to split Global and Local according to the bullet points from above

```suggestion
## Global Reporting Control
```

gogonzo commented on vignettes/managing-reproducible-report-documents-in-teal.rmd:
```diff
@@ -0,0 +1,195 @@
+custom_reporter_template <- Reporter$new()
+custom_reporter_template$set_template(
+  # Need to return named `teal_card`
+  function(card) {
```
Can you show some practical information, like setting a Rmd template or adding some header/footer etc.?

m7pr replied:
I was rather thinking about setting some templated text for disclaimers, data sources, data privacy information, etc as an information added on top or at the bottom of the card. This sentence can also be added to this example.

The `Rmd` template could be useful, if we create some text that can be interpreted as `rmarkdown`. Maybe including a logo that is taken from the web?

gogonzo commented on vignettes/managing-reproducible-report-documents-in-teal.rmd:
```diff
@@ -0,0 +1,195 @@
+
+#### Module-Level Reporting Control
+
+While the `reporter` parameter controls reporting globally, you can also disable reporting for specific modules using `disable_report()`. This allows you to have fine-grained control over which modules support reporting:
```
Both, `after` and `disable_report` can control module's report. After is more powerful and it doesn't restrict to the report only but to the whole `teal_report/teal_data` object. We can showcase how to modify the report using `after` 
 
```suggestion
While the `reporter` parameter controls reporting globally, you can also manage a report document for a specific module using `after()` `disable_report()`:
```

m7pr replied:
@gogonzo there is a section following this sentence that presents the usage of `after`:
`Using the `after` Function to Customize Module Report Cards`

Do you think it's enough? It shows that you can adjust the `teal_card` of `teal_report`, but we can apply `join_keys` modifications as well, and also an `eval_code` to show that the whole `teal_data/teal_report` can be changed in any direction, not only the `@teal_report` field.

m7pr commented on vignettes/adding-support-for-reporting.Rmd:
```diff
@@ -93,23 +93,24 @@ knitr::include_url(url, height = "800px")
   module(
     label = label,
-    server = function(id, data, reporter) {
+    server = function(id, data) {
```
However this very same code is added in the next section `Add content to the report card`

So maybe keep `Modify the body of the server function` as it is, and extend `Add content to the report card` to use `eval_code`

gogonzo replied:
I don't know if this vignette needs to have two examples. Previously adding reporter was more complicated now Reporter is supported by default as `teal_report` is passed to the teal_module's server. Only thing which user has to do is `return(<teal_report>)`. I'd just mention `keep_output` and add one markdown-heading element.
